### DISCUSSION TITLE
Streaming Support and Dynamic/Lazy Schemas

### DISCUSSION THREAD

**Discussion by rlhagen** [graphql-java/graphql-java#3183] [Ideas]
I work for a company that has implemented streaming support as well as something we call dynamic schemas.  The streaming support is probably self explanatory but I can elaborate if needed.  

The dynamic schema support is really lazy loading of schemas/fields at runtime vs loading up all the definitions when you build the GraphQL object.  So in short, we build out the fields lazily against a model that we have.  This saves a lot of time on startup. 

I'm reaching out to see if the maintainers of graphql-java are open to PRs for these features.

**Comment by bbakerman** [comment]
We would love to see the implementation ideas.

If you want to do that via PRs then great but if its easy to explain the general approach in words then thats great too.

if the PRs are easy to create then that might be a great way - eg put up a non complete PR say as a way to visualise code.  But words can work as well.

We are planning to add back in "@stream / @defer" back into graphql-java at some point - we had it but it was too early in a spec sense and the spec evolved and then didnt really move anywhere for a long time.

So we removed it until the specification firmed up.  I think that time of @stream / @defer is coming back as the spec now seems more stable.  That said we dont have a planned work item for that yet (graphql-java is volunteer based - not a pure company backed thing)

But we would love to see what you have done.

**Reply by rlhagen** [reply]
@bbakerman we also implemented the new RFC for stream and defer and are using it in production as well.  That might be useful as well.  I'll work on framing the features up with descriptions before I submit any PR since we'll need some time to to refactor those.

**Reply by rlhagen** [reply]
Streaming Support:  This feature enables the ability to write to an OutputStream rather than block and then write the entire response.  Our implementation overloads the `execute` method on the `GraphQL` class along with providing a `StreamingExecutionStrategy`.   So basically, there is away to pass the OutputStream from the HTTP response directly to GraphQL. 

Dynamic/Lazy Schemas:  This feature allows for lazy loading of the GraphQL definitions.  Rather than loading up concrete definitions, we added a `Provider` on the `GraphQLSchema` class that lazy loads fields and data fetchers.  We have a rather unique use case where we allow users to generate schemas on the fly at runtime as well as some schemas are generated from other user editable models from a database.  Without this feature, the build time for a GraphQLSchema was too large.  Also, only a tiny subset of the user editable models are actually used in the GraphQL so it saves a ton of memory (and we have no idea which ones will be used).  

Stream/Defer:  This is an implementation of the latest RFC (2022ish).  We can get this contributed if desired.  Might need a few tweaks since I haven't looked at the RFC in probably a year now.  But we implemented this in 2022.

**Comment by bbakerman** [comment]
By all means we would love to see this code.  I cant give you  a guarantee that we will accept it BUT we would love to see it.

Regarding lazy schemas - could you not use SchemaTransformer to make a new schema with the new elements when they change?

We would love to see your stream / defer implementation if possible.  At least for inspiration on what things you have in place

**Reply by rlhagen** [reply]
Alright.  I will be working on PRs when workload allows.  

I'd have to think about the SchemaTransformer and see how that would work.
